tests/integration/components/async-button-test.js

```diff
@@ -0,0 +1,63 @@
+import Ember from 'ember';
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent('async-button', 'Integration | Component | async button', {
+  integration: true
+});
+
+test('it responds to a fulfilled closure promise', function(assert) {
+  let done = assert.async();
+
+  assert.expect(2);
+
+  this.set('default', 'Save');
+  this.set('pending', 'Saving');
+  this.set('fulfilled', 'Saved');
+
+  this.on('closurePromise', function() {
+    return new Ember.RSVP.Promise((fulfill) => {
+      Ember.run.later(() => {
+        fulfill();
+      }, 500);
+    });
+  });
+
+  this.render(hbs`{{async-button default=default pending=pending fulfilled=fulfilled action=(action "closurePromise")}}`);
+  this.$('button').click();
+
+  assert.equal(this.$('button').text().trim(), 'Saving');
+
+  Ember.run.later(() => {
```

Comment:
A better pattern would be to import the `wait` helper from `ember-test-helpers`.  It is automatically aware of pending run loops (like from `Ember.run.next` or `Ember.run.later` above) and will resolve when they are completed.

This might look like:

```js
import wait from 'ember-test-helpers/wait';

// ....snip....

test('asdf', function(assert) {
  // other test stuff here

  wait()
    .then(() => {
      assert.equal(this.$('button').text().trim(), 'Saved');
      done();
    });
});
```

Also see: https://emberweekend.com/blog/851fb1e6f7-using-the-wait-helper-in-ember-integration-tests